### ISSUE TITLE
Possible memory leak?

### ISSUE THREAD

**Issue by b-ma** [orottier/web-audio-api-rs#45]
Hi,

First, thanks for your work on this promising project. 

I was playing with your library to create some Node.js bindings  (using [https://napi.rs/](https://napi.rs/)) and after some minimal work to wrap your code, I wrote this small javascript test. However the sound rapidly starts to be very noisy (a kind of nice noise but not the kind we would expect...), with the script rapidly growing to take almost 100% of a processor:

```js
function triggerSine(audioContext) {
  const sine = audioContext.createOscillator();
  sine.connect(audioContext.destination);

  const now = audioContext.currentTime;
  sine.start(now);
  sine.stop(now + 0.03);
}

(async function() {
  const audioContext = new AudioContext();
  await audioContext.resume();
  setInterval(() => triggerSine(audioContext), 0.05 * 1000);
}());
```

While I agree this may appear as a kind of "agressive" test, this kind of pattern is quite common e.g. for granular synthesis or for generative music.

To make sure, the problem does not come from my own code, I re-wrote the test in pure Rust but ran into the exact same issue:

```rust
use web_audio_api::context::{AsBaseAudioContext, AudioContext};
use web_audio_api::node::{
  AudioNode, AudioScheduledSourceNode, OscillatorType, PeriodicWaveOptions,
};
use std::{thread, time};

fn trigger_sine(audio_context: &AudioContext) {
  let osc = audio_context.create_oscillator();
  osc.connect(&audio_context.destination());

  let now = audio_context.current_time();
  osc.start_at(now);
  osc.stop_at(now + 0.03)
}

fn main() {
  let audio_context = AudioContext::new();

  // mimic setInterval
  loop {
    trigger_sine(&audio_context);
    thread::sleep(time::Duration::from_millis(50));
  }
}
```

As I'm a newbie with Rust, I didn't try to dig more deeply into your code, but I would be happy to help the best I can.

Cheers,
Benjamin

**Comment by orottier** [comment]
Hi @b-ma, thanks for the detailed report. It is very helpful.

I believe your assumptions are correct and your rust code should not cause a memory leak and/or performance issue.
My guess is that we currently do not clean up oscillator nodes from the audio graph as they are stopped. Therefore the audio graph will continue growing and use up memory and CPU.

Specifically at https://github.com/orottier/web-audio-api-rs/blob/main/src/node/oscillator.rs#L739 we set tail-time `true` always, preventing deallocation of the node. This should be changed to something like `return self.scheduler.is_stopped()`. The spec indicates that stopped nodes cannot be started again so that would be compliant

**Comment by orottier** [comment]
I added a work-in-progress PR to improve the OscillatorNode cleanup of inactive nodes, but the runaway memory and CPU issue persists. To be continued...

**Comment by orottier** [comment]
Fixed in #46

**Comment by b-ma** [comment]
cool !
